crates/corro-agent/src/agent/util.rs

```diff
@@ -0,0 +1,1479 @@
+                tripwire
+                    .clone()
+                    .inspect(move |_| info!("corrosion api http tripped {api_addr}")),
+            )
+            .inspect(|_| info!("corrosion api is done")),
+    );
+
+    Ok(())
+}
+
+async fn require_authz<B>(
+    Extension(agent): Extension<Agent>,
+    maybe_authz_header: Option<TypedHeader<Authorization<Bearer>>>,
+    request: axum::http::Request<B>,
+    next: axum::middleware::Next<B>,
+) -> Result<axum::response::Response, axum::http::StatusCode> {
+    let passed = if let Some(ref authz) = agent.config().api.authorization {
+        match authz {
+            AuthzConfig::BearerToken(token) => maybe_authz_header
+                .map(|h| h.token() == token)
+                .unwrap_or(false),
+        }
+    } else {
+        true
+    };
+
+    if !passed {
+        return Err(axum::http::StatusCode::UNAUTHORIZED);
+    }
+
+    Ok(next.run(request).await)
+}
+
+// DOCME: provide some context for this function
+// TODO: move to a more appropriate module?
+#[tracing::instrument(skip_all)]
+pub fn find_cleared_db_versions(
+    tx: &Transaction,
+    actor_id: &ActorId,
+) -> rusqlite::Result<BTreeSet<CrsqlDbVersion>> {
+    let clock_site_id: Option<u64> = match tx
+        .prepare_cached("SELECT ordinal FROM crsql_site_id WHERE site_id = ?")?
+        .query_row([actor_id], |row| row.get(0))
+        .optional()?
+    {
+        Some(0) => None, // this is the current crsql_site_id which is going to be NULL in clock tables
+        Some(ordinal) => Some(ordinal),
+        None => {
+            warn!(actor_id = %actor_id, "could not find crsql ordinal for actor");
+            return Ok(Default::default());
+        }
+    };
+
+    let tables = tx
+        .prepare_cached(
+            "SELECT name FROM sqlite_schema WHERE type = 'table' AND name LIKE '%__crsql_clock'",
+        )?
+        .query_map([], |row| row.get(0))?
+        .collect::<Result<BTreeSet<String>, _>>()?;
+
+    if tables.is_empty() {
+        // means there's no schema trakced by cr-sqlite or corrosion.
+        return Ok(BTreeSet::new());
+    }
+
+    let mut params: Vec<&dyn ToSql> = vec![actor_id];
+    let to_clear_query = format!(
+        "SELECT DISTINCT db_version FROM __corro_bookkeeping WHERE actor_id = ? AND db_version IS NOT NULL
+            EXCEPT SELECT db_version FROM ({});",
+        tables
+            .iter()
+            .map(|table| {
+                params.push(&clock_site_id);
+                format!("SELECT DISTINCT db_version FROM {table} WHERE site_id IS ?")
+            })
+            .collect::<Vec<_>>()
+            .join(" UNION ")
+    );
+
+    let cleared_db_versions: BTreeSet<CrsqlDbVersion> = tx
+        .prepare_cached(&to_clear_query)?
+        .query_map(params_from_iter(params.into_iter()), |row| row.get(0))?
+        .collect::<rusqlite::Result<_>>()?;
+
+    Ok(cleared_db_versions)
+}
+
+// DOCME: provide some context for this function
```

Comment:
This is a loop that handles periodically initiating a sync from the current node to many other nodes (via another function), but also applying "buffered" (partial) changesets (like partial transactions).

It works this way so we try and process these buffered changes before we try to sync so we have less to sync.